### PR TITLE
Enable open tracing propagation in workflow lifecycles

### PR DESCRIPTION
**Pull request by shijiesheng**
**How it works**
Context Propagation in Cadence (Customer)

On start workflow API, trace span with context is written into workflow start event attributes, which is persisted in cadence server side. 
On workflow-start in client, this span is referenced and activated on execute workflow. 
On scheduling child workflows and activities (including local activities), the span is written into child workflow's workflow start event attributes and activity's schedule activity event attributes. 
On processing activities/childworkflows, the persisted span is referenced and activated again. 

**Sample Spans**

Notes: Poll + Respond apis spans are omitted here

{traceId:1, spanId:2, parentId:0, operationName:"cadence-RegisterDomain"}
{traceId:1, spanId:3, parentId:2, operationName:"Test Started"}
{traceId:1, spanId:18, parentId:3, operationName:"cadence-StartWorkflowExecution"}
{traceId:1, spanId:19, parentId:18, operationName:"cadence-GetWorkflowExecutionHistory"}
{traceId:1, spanId:21, parentId:18, operationName:"cadence-ExecuteWorkflow"}
{traceId:1, spanId:24, parentId:21, operationName:"cadence-ExecuteActivity"}
{traceId:1, spanId:25, parentId:24, operationName:"cadence-RespondActivityTaskCompleted"}
{traceId:1, spanId:31, parentId:21, operationName:"cadence-ExecuteWorkflow"}
{traceId:1, spanId:32, parentId:31, operationName:"cadence-ExecuteLocalActivity"}

**What changed?**

* added an Propagator entity with tracing extract/inject logic
* added trace activation logic in activity and workflow executors
* added trace activation on service client (Tchannel + GRPC)

**Why?**

improve observability

**How did you test it?**

integration test

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: src/main/java/com/uber/cadence/serviceclient/WorkflowServiceTChannel.java [hunk 1]
```diff
@@ -339,6 +345,8 @@ private <T> T measureRemoteCallWithTags(
       sw.stop();
       scope.counter(MetricsType.CADENCE_ERROR).inc(1);
       throw e;
+    } finally {
+      span.finish();
```
Comment (davidporter-id-au) [inline src/main/java/com/uber/cadence/serviceclient/WorkflowServiceTChannel.java:349]: out of curiuosity, is there any kind of error attach you can put on the span to indicate the type of error? I vaguely seem to recall that was possible

Comment (davidporter-id-au) [reply src/main/java/com/uber/cadence/serviceclient/WorkflowServiceTChannel.java:349]: ah, it's a tag: https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table

File: src/main/java/com/uber/cadence/internal/sync/SyncDecisionContext.java [hunk 2]
```diff
@@ -130,9 +154,18 @@ public WorkflowInterceptor getWorkflowInterceptor() {
   @Override
   public byte[] executeWorkflow(
       SyncWorkflowDefinition workflowDefinition, WorkflowExecuteInput input) {
-    return workflowDefinition.execute(input.getInput());
+    Span span = tracingPropagator.activateSpanForExecuteWorkflow(context);
+    try {
+      return workflowDefinition.execute(input.getInput());
+    } finally {
+      span.finish();
```
Comment (davidporter-id-au) [inline src/main/java/com/uber/cadence/internal/sync/SyncDecisionContext.java:161]: also as before, it *might* be worth checking if there was an error (to the extent I understand what's happening)

**Review by davidporter-id-au — APPROVED** [review]
I'm curious about some of the architectural pieces here, how we'll represent very long running workflows in the context of a Cadence workflow, but otherwise lgtm!

File: src/main/java/com/uber/cadence/serviceclient/WorkflowServiceTChannel.java (hunk 1 above)
Comment (shijiesheng) [reply src/main/java/com/uber/cadence/serviceclient/WorkflowServiceTChannel.java:349]: Yes, we can also add logs.

**Comment by shijiesheng** [comment]
> I'm curious about some of the architectural pieces here, how we'll represent very long running workflows in the context of a Cadence workflow, but otherwise lgtm!

For jaeger, that has a TTL set for each span. In long running workflows, only recent spans will exist and the DAG view will be constructed by the remaining ones. These spans still have references to parents, but these parents will not be found.

File: src/main/java/com/uber/cadence/internal/sync/SyncDecisionContext.java (hunk 2 above)
Comment (shijiesheng) [reply src/main/java/com/uber/cadence/internal/sync/SyncDecisionContext.java:161]: I'll add that in a later PR

**Pull request merged**